pdns/syncres.cc

```diff
@@ -1282,6 +1282,36 @@ inline vector<DNSName> SyncRes::shuffleInSpeedOrder(NsSet &tnameservers, const s
   return rnameservers;
 }
 
+inline vector<ComboAddress> SyncRes::shuffleForwardSpeed(vector<ComboAddress> &rnameservers, const string &prefix, const bool wasRd)
```

Comment:
Since we are going to copy the whole vector anyway, perhaps you could do the copy first and do the random shuffle and stable sort on the copy, allowing `rnameservers` to be `const`?

Reply:
Like this, or do you mean something else?

```
diff --git a/pdns/syncres.cc b/pdns/syncres.cc
index 67b888228..e39ae1799 100644
--- a/pdns/syncres.cc
+++ b/pdns/syncres.cc
@@ -1282,26 +1282,27 @@ inline vector<DNSName> SyncRes::shuffleInSpeedOrder(NsSet &tnameservers, const s
   return rnameservers;
 }

-inline vector<ComboAddress> SyncRes::shuffleForwardSpeed(vector<ComboAddress> &rnameservers, const string &prefix, const bool wasRd)
+inline vector<ComboAddress> SyncRes::shuffleForwardSpeed(const vector<ComboAddress> &rnameservers, const string &prefix, const bool wasRd)
 {
+  vector<ComboAddress> nameservers = rnameservers;
   map<ComboAddress, double> speeds;

-  for(const auto& val: rnameservers) {
+  for(const auto& val: nameservers) {
     double speed;
     DNSName nsName = DNSName(val.toStringWithPort());
     speed=t_sstorage.nsSpeeds[nsName].get(&d_now);
     speeds[val]=speed;
   }
-  random_shuffle(rnameservers.begin(),rnameservers.end(), dns_random);
+  random_shuffle(nameservers.begin(),nameservers.end(), dns_random);
   speedOrderCA so(speeds);
-  stable_sort(rnameservers.begin(),rnameservers.end(), so);
+  stable_sort(nameservers.begin(),nameservers.end(), so);

   if(doLog()) {
     LOG(prefix<<"Nameservers: ");
-    for(vector<ComboAddress>::const_iterator i=rnameservers.cbegin();i!=rnameservers.cend();++i) {
-      if(i!=rnameservers.cbegin()) {
+    for(vector<ComboAddress>::const_iterator i=nameservers.cbegin();i!=nameservers.cend();++i) {
+      if(i!=nameservers.cbegin()) {
         LOG(", ");
-        if(!((i-rnameservers.cbegin())%3)) {
+        if(!((i-nameservers.cbegin())%3)) {
           LOG(endl<<prefix<<"             ");
         }
       }
@@ -1309,7 +1310,7 @@ inline vector<ComboAddress> SyncRes::shuffleForwardSpeed(vector<ComboAddress> &r
     }
     LOG(endl);
   }
-  return rnameservers;
+  return nameservers;
 }

 static uint32_t getRRSIGTTL(const time_t now, const std::shared_ptr<RRSIGRecordContent>& rrsig)
diff --git a/pdns/syncres.hh b/pdns/syncres.hh
index 702c27486..738d18fa7 100644
--- a/pdns/syncres.hh
+++ b/pdns/syncres.hh
@@ -739,7 +739,7 @@ private:
   DNSName getBestNSNamesFromCache(const DNSName &qname, const QType &qtype, NsSet& nsset, bool* flawedNSSet, unsigned int depth, set<GetBestNSAnswer>&beenthere);

   inline vector<DNSName> shuffleInSpeedOrder(NsSet &nameservers, const string &prefix);
-  inline vector<ComboAddress> shuffleForwardSpeed(vector<ComboAddress> &rnameservers, const string &prefix, const bool wasRd);
+  inline vector<ComboAddress> shuffleForwardSpeed(const vector<ComboAddress> &rnameservers, const string &prefix, const bool wasRd);
   bool moreSpecificThan(const DNSName& a, const DNSName &b) const;
   vector<ComboAddress> getAddrs(const DNSName &qname, unsigned int depth, set<GetBestNSAnswer>& beenthere, bool cacheOnly);

```

Reply:
Yes, that's exactly what I had in mind! Not altering our parameters makes reviewing the side effects of this method so much easier :)